### PR TITLE
Add table showing complexity of methods for Input

### LINKED ISSUE #10106: Document computational complexity of `bevy::input::Input`
The documentation for [`Query`](https://docs.rs/bevy/latest/bevy/ecs/system/struct.Query.html#performance) has a table about the computational complexity of its APIs. Such a table would be helpful, and I'd like to see this added to the documentation for [`bevy::input::Input`](https://docs.rs/bevy/latest/bevy/input/struct.Input.html).

The table in the `Query` documentation contains APIs related to iteration (`iter`, `par_iter`, etc.), querying (`get`, `single`, etc.), and filtering (`With`, `Added`, etc.).

If a similar table is added to `Input`, I'd like to see it describe at least the following APIs. (as of Bevy 0.11.3)

- [`any_just_pressed`](https://docs.rs/bevy/0.11.3/bevy/input/struct.Input.html#method.any_just_pressed)
- [`any_just_released`](https://docs.rs/bevy/0.11.3/bevy/input/struct.Input.html#method.any_just_released)
- [`any_pressed`](https://docs.rs/bevy/0.11.3/bevy/input/struct.Input.html#method.any_pressed)
- [`get_just_pressed`](https://docs.rs/bevy/0.11.3/bevy/input/struct.Input.html#method.get_just_pressed)
- [`get_just_released`](https://docs.rs/bevy/0.11.3/bevy/input/struct.Input.html#method.get_just_released)
- [`get_pressed`](https://docs.rs/bevy/0.11.3/bevy/input/struct.Input.html#method.get_pressed)
- [`just_pressed`](https://docs.rs/bevy/0.11.3/bevy/input/struct.Input.html#method.just_pressed)
- [`just_released`](https://docs.rs/bevy/0.11.3/bevy/input/struct.Input.html#method.just_released)
- [`pressed`](https://docs.rs/bevy/0.11.3/bevy/input/struct.Input.html#method.pressed)

### PR DESCRIPTION
**Pull request by ShadowMitia**
# Objective

Fixes #10106 

Adds a table showing computational complexity, as seen for Query (and similar to std::collections docs).

## Solution

Add the complexity table

---

## Changelog

- Add complexity table for Input methods

### CODE REVIEW

**Comment by ShadowMitia** [comment]
I've added information on all methods, hope this is ok.

I've tried analyzing the source and I've also used the [std::collections doc](https://doc.rust-lang.org/stable/std/collections/index.html) on performance also as a reference.

I'm not completely sure on the values for just_pressed/just_released/pressed, which all have a call to `contain`.
I'm kind of assuming it is comparable to a `get` complexity as seen in the collections docs.

File: crates/bevy_input/src/input.rs [hunk 1]
```diff
@@ -29,6 +29,35 @@ use bevy_ecs::schedule::State;
...
+/// - **n** is the number of stored inputs.
+/// - **m** is the number of elements of another collection involved in the operation.
+/// - **\***-suffix denotes an amortized cost.
+/// - **~**-suffix denotes an expected cost.
+///
+/// See Rust's [std::collections doc on performance](https://doc.rust-lang.org/std/collections/index.html#performance) for more details on the conventions used here.
+///
+/// | **[`Input`] operations**          | **Computational complexity**       |
+/// |-----------------------------------|------------------------------------|
+/// | [`Input::any_just_pressed`]       | *O*(m*n)                           |
```
Comment (alice-i-cecile) [inline crates/bevy_input/src/input.rs:44]: I think n * m will be slightly easier to read.

File: crates/bevy_input/src/input.rs [hunk 2]
```diff
@@ -29,6 +29,35 @@ use bevy_ecs::schedule::State;
 /// * Using [`Input::clear_just_pressed`] or [`Input::clear_just_released`] instead.
 /// * Calling [`Input::clear`] or [`Input::reset`] immediately after the state change.
 ///
+/// ## Performance
+///
+/// For all operations, the following conventions are used:
+/// - **n** is the number of stored inputs.
+/// - **m** is the number of elements of another collection involved in the operation.
+/// - **\***-suffix denotes an amortized cost.
```
Comment (alice-i-cecile) [inline crates/bevy_input/src/input.rs:37]: Given that this isn't used here, I think it should be cut from this table.

**Review by alice-i-cecile — APPROVED** [review]
Nits on the wording, but nothing blocking. All of the listed time complexities appear correct to me.

File: crates/bevy_input/src/input.rs (hunk 2 above)
Comment (ShadowMitia) [reply crates/bevy_input/src/input.rs:37]: Sorry I don't understand, I'm using all the symbols?
To be sure, this **\*** is for the * symbol.

Comment (alice-i-cecile) [reply crates/bevy_input/src/input.rs:37]: My bad, I see now. It was hard to follow in markdown.

File: crates/bevy_input/src/input.rs [hunk 3]
```diff
@@ -29,6 +29,35 @@ use bevy_ecs::schedule::State;
...
+/// | [`Input::just_pressed`]           | *O*(1)~                            |
+/// | [`Input::just_released`]          | *O*(1)~                            |
+/// | [`Input::pressed`]                | *O*(1)~                            |
+/// | [`Input::press`]                  | *O*(1)~*                           |
+/// | [`Input::release`]                | *O*(1)~*                           |
+/// | [`Input::release_all`]            | *O*(n)~*                           |
+/// | [`Input::clear_just_pressed`]     | *O*(1)~                            |
+/// | [`Input::clear_just_released`]    | *O*(1)~                            |
+/// | [`Input::reset_all`]              | *O*(n)                             |
+/// | [`Input::clear`]                  | *O*(n)                             |
```
Comment (james7132) [inline crates/bevy_input/src/input.rs:59]: Is this accurate? HashSet::clear is O(1), no?

Comment (ShadowMitia) [reply crates/bevy_input/src/input.rs:59]: I'll double-check properly when I can, and I could be wrong, but on clear the hashtable goes through all elements to drop them one at a time.
Maybe this is only true for types with a custom drop?

Comment (ShadowMitia) [reply crates/bevy_input/src/input.rs:59]: HashSets are based on HashMaps, and both use an implementation based on the hasbrown crate.
If you follow down the clear() to where it is actually doing things, you get here:
https://github.com/rust-lang/hashbrown/blob/f2e62124cd947b5e2309dd6a24c7e422932aae97/src/raw/mod.rs#L1091
Which doesn't do much except call drop_elements which is defined here:
https://github.com/rust-lang/hashbrown/blob/f2e62124cd947b5e2309dd6a24c7e422932aae97/src/raw/mod.rs#L2369
That last one I consider to be `O(n)` hence the information put into the table. But its only run for cases where the type needs to be dropped.

So I guess I need to specify both cases? Or should we say it is `O(n)` worse case?

Comment (matiqo15) [reply crates/bevy_input/src/input.rs:59]: I would go for `O(n)` because it's standard to refer to the worst case scenario when talking about time complexity.

File: crates/bevy_input/src/input.rs [hunk 4]
```diff
@@ -29,6 +29,35 @@ use bevy_ecs::schedule::State;
 /// * Using [`Input::clear_just_pressed`] or [`Input::clear_just_released`] instead.
 /// * Calling [`Input::clear`] or [`Input::reset`] immediately after the state change.
 ///
+/// ## Performance
+///
+/// For all operations, the following conventions are used:
+/// - **n** is the number of stored inputs.
+/// - **m** is the number of input types.
```
Comment (mockersf) [inline crates/bevy_input/src/input.rs:36]: I'm not sure of the meaning of "input types", there's only one type `T`

Comment (nelsontkq) [reply crates/bevy_input/src/input.rs:36]:
```suggestion
/// - **m** is the number of input parameters passed to the method.
```

File: crates/bevy_input/src/input.rs (hunk 3 above)
Comment (nelsontkq) [reply crates/bevy_input/src/input.rs:59]: Looks like [it's marked as O(1)~ in the std docs](https://doc.rust-lang.org/stable/std/collections/index.html#maps).

I think it's more useful to know it's O(1) in 99.999999% of cases, and most likely in this case since there's probably less than 20 things in each of those sets.

**Comment by ShadowMitia** [comment]
Updated with all suggestions and rebased on latest changes.
I still think the clear method should be documented as `O(n)`, unless there is a good reason not to :smile: 
Should be good to go otherwise!

**Review by james7132 — APPROVED** [review]
